reassembly_queue.go

```diff
@@ -1,116 +1,329 @@
 
-	firstPacket := m.fragmentQueue[0]
-	if len(m.fragmentQueue) == 1 {
-		return firstPacket.beginingFragment && firstPacket.endingFragment
+	if chunk.unordered {
+		// First, insert into unorderedChunks array
+		r.unorderedChunks = append(r.unorderedChunks, chunk)
+		sortChunksByTSN(r.unorderedChunks)
+
+		// Scan unorderedChunks that are contiguous (in TSN)
+		cset = r.findCompleteUnorderedChunkSet()
+
+		// If found, append the complete set to the unordered array
+		if cset != nil {
+			r.unordered = append(r.unordered, cset)
+			return true
+		}
+
+		return false
 	}
 
-	lastPacket := m.fragmentQueue[len(m.fragmentQueue)-1]
-	return firstPacket.beginingFragment && lastPacket.endingFragment
+	// This is an ordered chunk
 
-}
+	//fmt.Printf("SI[%d]: push SSN %d Next SSN=%d\n", r.si, chunk.streamSequenceNumber, r.nextSSN)
+
+	if sna16LT(chunk.streamSequenceNumber, r.nextSSN) {
+		fmt.Printf("SI[%d]: ignore stale SSN %d < %d\n", r.si, chunk.streamSequenceNumber, r.nextSSN)
+		return false
+	}
+
+	// Check if a chunkSet with the SSN already exists
+	for _, set := range r.ordered {
+		if set.ssn == chunk.streamSequenceNumber {
+			cset = set
+			break
+		}
+	}
 
-func (m *payloadDataMessage) clear() {
-	m.length = 0
-	m.fragmentQueue = []*chunkPayloadData{}
+	// If not found, create a new chunkSet
+	if cset == nil {
+		cset = newChunkSet(chunk.streamSequenceNumber, chunk.payloadType)
+		r.ordered = append(r.ordered, cset)
+		if !chunk.unordered {
+			sortChunksBySSN(r.ordered)
+		}
+	}
+
+	return cset.push(chunk)
 }
 
-func (m *payloadDataMessage) assemble() ([]byte, bool) {
-	if m.complete() {
-		b := make([]byte, m.length)
-		i := 0
-		for _, p := range m.fragmentQueue {
-			copy(b[i:], p.userData)
-			i += len(p.userData)
+func (r *reassemblyQueue) findCompleteUnorderedChunkSet() *chunkSet {
+	startIdx := -1
+	nChunks := 0
+	var lastTSN uint32
+	var found bool
+
+	for i, c := range r.unorderedChunks {
+		// seek beigining
+		if c.beginningFragment {
+			startIdx = i
+			nChunks = 1
+			lastTSN = c.tsn
+
+			if c.endingFragment {
+				found = true
+				break
+			}
+			continue
+		}
+
+		if startIdx < 0 {
+			continue
+		}
+
+		// Check if contiguous in TSN
+		if c.tsn != lastTSN+1 {
+			startIdx = -1
+			continue
+		}
+
+		lastTSN = c.tsn
+		nChunks++
+
+		if c.endingFragment {
+			found = true
+			break
 		}
+	}
 
-		return b, true
+	if !found {
+		return nil
 	}
 
-	return nil, false
+	// Extract the range of chunks
+	chunks := r.unorderedChunks[startIdx : startIdx+nChunks]
+	r.unorderedChunks = append(
+		r.unorderedChunks[:startIdx],
+		r.unorderedChunks[startIdx+nChunks:]...)
+
+	chunkSet := newChunkSet(0, chunks[0].payloadType)
+	for _, c := range chunks {
+		chunkSet.push(c)
+	}
+
+	return chunkSet
 }
 
-type reassemblyQueue struct {
-	messageQueue     payloadDataMessageArray
-	unorderedMessage payloadDataMessage
-	expectedSeqNum   uint16
+func (r *reassemblyQueue) isReadable() bool {
+	// Check unordered first
+	if len(r.unordered) > 0 {
```

Comment:
Does it make sense the mark either of these as readable when packets are added to fast-exit?

Reply:
I was debating on that... If we introduce an internal flag, "readable", then it has to be maintained by many places, where the current one simplifies it with a bit of cost of evaluating the arrays...

Reply:
I'm totally fine with that answer! Caching things should be a post-evaluation, hotspot squashing activity